src/clients/SpokePoolClient.ts

```diff
@@ -43,256 +42,271 @@ export function isSpokePoolEventRemoved(message: unknown): message is SpokePoolE
-      pendingEvents.splice(pendingEventIdx, 1);
+        pendingEvents.forEach((event) => {
+          const eventIdx = this._queryableEventNames().indexOf(event.event);
+          assert(
+            eventIdx !== -1 && event.removed === false,
+            event.removed ? "Incorrectly received removed event" : `Unsupported event name (${event.event})`
+          );
 
-      this.logger.debug({
-        at: "SpokePoolClient#removeEvent",
-        message: `Removed ${this.chain} ${eventName} event for block ${blockNumber}.`,
-        event,
-      });
+          this.#pendingEvents[eventIdx].push(event);
+        });
+      }
+
+      this.#pendingBlockNumber = blockNumber;
+      this.#pendingCurrentTime = currentTime;
     }
 
-    // Back out any events that were previously ingested via update(). This is best-effort and may help to save the
-    // relayer from filling a deposit where it must wait for additional deposit confirmations. Note that this is
-    // _unsafe_ to do ad-hoc, since it may interfere with some ongoing relayer computations relying on the
-    // depositHashes object. If that's an acceptable risk then it might be preferable to simply assert().
-    if (eventName === "V3FundsDeposited" || eventName === "FundsDeposited") {
-      const { depositId } = event.args;
-      assert(isDefined(depositId));
-
-      const depositEvent = {
-        ...spreadEventWithBlockNumber(event),
-        messageHash: event.args.messageHash ?? getMessageHash(event.args.message),
-      } as DepositWithBlock;
-      const depositHash = getRelayEventKey(depositEvent);
-      if (isDefined(this.depositHashes[depositHash])) {
-        delete this.depositHashes[depositHash];
-        this.logger.warn({
+    /**
+     * Given an event to be removed, ensure that it is removed from the set of ingested events.
+     * @param event An Ethers event instance.
+     * @returns void
+     */
+    #removeEvent(event: Log): boolean {
+      let removed = false;
+      const eventIdx = this._queryableEventNames().indexOf(event.event);
+      const pendingEvents = this.#pendingEvents[eventIdx];
+
+      const { event: eventName, blockNumber, blockHash, transactionHash, transactionIndex, logIndex } = event;
+
+      // First check for removal from any pending events.
+      const pendingEventIdx = pendingEvents.findIndex(
+        (pending) =>
+          pending.logIndex === logIndex &&
+          pending.transactionIndex === transactionIndex &&
+          pending.transactionHash === transactionHash &&
+          pending.blockHash === blockHash
+      );
+
+      if (pendingEventIdx !== -1) {
+        removed = true;
+
+        // Drop the relevant event.
+        pendingEvents.splice(pendingEventIdx, 1);
+
+        this.logger.debug({
           at: "SpokePoolClient#removeEvent",
-          message: `Removed 1 pre-ingested ${this.chain} ${eventName} event.`,
+          message: `Removed ${this.#chain} ${eventName} event for block ${blockNumber}.`,
```

Comment:
Did not intend to drop this.

```suggestion
          message: `Removed 1 pre-ingested ${this.#chain} ${eventName} event for block ${blockNumber}.`,
```